core/src/main/java/jenkins/install/SetupWizard.java

```diff
@@ -0,0 +1,259 @@
+                if (cookie == null || !randomUUID.equals(cookie.getValue())) {
+                    if (randomUUID.equals(req.getParameter(securityTokenParameterName))) {
+                        cookie = new Cookie(securityTokenParameterName, randomUUID);
+                        cookie.setPath(req.getContextPath());
+                        ((HttpServletResponse)response).addCookie(cookie);
+                    }
+                    else if (!Pattern.compile(".*[.](css|ttf|gif|woff|eot|png|js)").matcher(req.getRequestURI()).matches()) {
+                        // Allow js & css requests through
+                        chain.doFilter(new HttpServletRequestWrapper(req) {
+                            public String getRequestURI() {
+                                return getContextPath() + "/enter-security-token";
+                            }
+                        }, response);
+                        return;
+                    }
+                    // fall through to handling the request normally
+                }
+            }
+            chain.doFilter(request, response);
+        }
+
+        @Override
+        public void destroy() {
+        }
+    };
+
+    private final Logger LOGGER = Logger.getLogger(Jenkins.class.getName());
+    private String randomUUID;
+    private AssetManager assetManager = new AssetManager();
+    private I18n i18n = new I18n();
+    private SetupWizardSecurityConfiguration securityConfiguration = new SetupWizardSecurityConfiguration();
+
+    /**
+     * Whether this system requires a security token; -Djenkins.install.skip.security=true will
+     * allow skipping of this security token
+     */
+    private boolean requireSecurityToken = !"true".equalsIgnoreCase(System.getProperty("jenkins.install.skip.security"));
+
+    /**
+     * The security token parameter name
+     */
+    private String securityTokenParameterName = ".installWizardSecurityToken";
+
+    /**
+     * Constructor, will generate a security token
+     */
+    public SetupWizard() throws IOException, InterruptedException, ReactorException {
+        generateToken();
+    }
+
+    public String getSecurityTokenParameterName() {
+        return securityTokenParameterName;
+    }
+
+    @Nonnull
+    @Restricted(NoExternalUse.class)
+    Jenkins getInstance() { // not public so Stapler won't dispatch it
+        return Jenkins.getActiveInstance();
+    }
+
+    private void generateToken() {
+        if (requireSecurityToken) {
+            // if Hudson is newly set up with the security realm and there's no user account created yet,
+            // insert a filter that asks the user to create one
+            try {
+                PluginServletFilter.addFilter(VALIDATE_TOKEN_FILTER);
+            } catch (ServletException e) {
+                throw new AssertionError(e); // never happen because our Filter.init is no-op
+            }
+
+            randomUUID = UUID.randomUUID().toString().replace("-", "").toLowerCase();
```

Comment:
I think `UUID` uses only A–F but to be safe specify `Locale.ENGLISH`. (In Turkish, I → ı.)